src/common/s3.connection.cpp

```diff
@@ -318,8 +223,13 @@ common::S3ConnectionPool::S3ConnectionPool(size_t n_connections,
     CHECK(n_connections > 0);
 
     for (auto i = 0; i < n_connections; ++i) {
-        connections_.push_back(std::make_unique<S3Connection>(
-          endpoint, access_key_id, secret_access_key));
+        auto connection = std::make_unique<S3Connection>(
+          endpoint, access_key_id, secret_access_key);
+
+        if (connection->check_connection()) {
```

Comment:
- I see that you're checking if connections exist in `get_connection()`. Would it be better to raise an error in this method if we are unable to create active connections?
- Should this be an asynchronous operation? I think it would be nice to run this asynchronously using `std::async` and block in `get_connection()` until it’s done.